tests/functional_small/ops/ops_eigen_level3.cc

```diff
@@ -2,157 +2,238 @@
-  EXPECT_DOUBLE_EQ(myR(0,0), 0.0); \
-  EXPECT_DOUBLE_EQ(myR(0,1), 5.0); \
-  EXPECT_DOUBLE_EQ(myR(0,2), 10.0); \
-  EXPECT_DOUBLE_EQ(myR(1,0), 0.0); \
-  EXPECT_DOUBLE_EQ(myR(1,1), 3.0); \
-  EXPECT_DOUBLE_EQ(myR(1,2), 6.0); \
-  EXPECT_DOUBLE_EQ(myR(2,0), 0.0); \
-  EXPECT_DOUBLE_EQ(myR(2,1), 8.0); \
-  EXPECT_DOUBLE_EQ(myR(2,2), 16.0); \
-  /* also cover non-zero beta */ \
-  constexpr auto beta1 = ::pressio::utils::Constants<double>::one(); \
-  pressio::ops::product(pressio::transpose(), pressio::nontranspose(), \
-      alpha, M, OPERAND, beta1, myR); \
-  EXPECT_DOUBLE_EQ(myR(0,0), 0.0); \
-  EXPECT_DOUBLE_EQ(myR(0,1), 10.0); \
-  EXPECT_DOUBLE_EQ(myR(0,2), 20.0); \
-  EXPECT_DOUBLE_EQ(myR(1,0), 0.0); \
-  EXPECT_DOUBLE_EQ(myR(1,1), 6.0); \
-  EXPECT_DOUBLE_EQ(myR(1,2), 12.0); \
-  EXPECT_DOUBLE_EQ(myR(2,0), 0.0); \
-  EXPECT_DOUBLE_EQ(myR(2,1), 16.0); \
-  EXPECT_DOUBLE_EQ(myR(2,2), 32.0); \
-
-
-#define OPS_EIGEN_DENSE_MAT_T_SELF_PROD \
-  mat_t M(4,3);                     \
-  M << 1,0,2, 2,1,3, 0,0,1, 2,2,2;\
-                                  \
-  mat_t myR(3,3);         \
-  myR(0, 0) = std::nan("0"); /* simulate uninitialized NaN */ \
-  constexpr auto beta  = ::pressio::utils::Constants<double>::zero(); \
-  constexpr auto alpha = ::pressio::utils::Constants<double>::one();  \
-  pressio::ops::product(pressio::transpose(), pressio::nontranspose(), alpha, M, beta, myR); \
-  EXPECT_DOUBLE_EQ(myR(0,0), 9.0); \
-  EXPECT_DOUBLE_EQ(myR(0,1), 6.0); \
-  EXPECT_DOUBLE_EQ(myR(0,2), 12.0); \
-  EXPECT_DOUBLE_EQ(myR(1,0), 6.0); \
-  EXPECT_DOUBLE_EQ(myR(1,1), 5.0); \
-  EXPECT_DOUBLE_EQ(myR(1,2), 7.0); \
-  EXPECT_DOUBLE_EQ(myR(2,0), 12.0); \
-  EXPECT_DOUBLE_EQ(myR(2,1), 7.0); \
-  EXPECT_DOUBLE_EQ(myR(2,2), 18.0); \
-  /* also cover non-zero beta */ \
-  constexpr auto beta1 = ::pressio::utils::Constants<double>::one(); \
-  pressio::ops::product(pressio::transpose(), pressio::nontranspose(), alpha, M, beta1, myR); \
-  EXPECT_DOUBLE_EQ(myR(0,0), 18.0); \
-  EXPECT_DOUBLE_EQ(myR(0,1), 12.0); \
-  EXPECT_DOUBLE_EQ(myR(0,2), 24.0); \
-  EXPECT_DOUBLE_EQ(myR(1,0), 12.0); \
-  EXPECT_DOUBLE_EQ(myR(1,1), 10.0); \
-  EXPECT_DOUBLE_EQ(myR(1,2), 14.0); \
-  EXPECT_DOUBLE_EQ(myR(2,0), 24.0); \
-  EXPECT_DOUBLE_EQ(myR(2,1), 14.0); \
-  EXPECT_DOUBLE_EQ(myR(2,2), 36.0); \
-
-
-namespace {
-template<class T>
-void fillOperand1(T & M)
+//-------------------------------------------
+// work around weird ::pressio::as_diagonal_matrix semantics
+// which does not produce functional matrix
+//-------------------------------------------
+
+template <typename VecType>
+class ExprAdapter {
+public:
+  using diagexpr_t = pressio::expressions::impl::AsDiagonalMatrixExpr<VecType>;
+  using sc_t = typename ::pressio::Traits<diagexpr_t>::scalar_type;
+
+  ExprAdapter(const diagexpr_t &expr): expr_(expr) {}
+
+  sc_t operator()(size_t i, size_t j) const {
+    return i == j ? expr_(i, j) : ::pressio::utils::Constants<sc_t>::zero();
+  }
```

Comment:
@fnrizzi
What's the semantics of objects returned by `::pressio::as_diagonal_matrix` ?
Do you think they could act as read-only matrices and yield zeros for non-diagonal elements ?

Reply:
it should be same semantics as a matrix, so we can also overwrite values
but for all non-diag elements it should return zero

Reply:
thanks, created #532 to track this